A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,516 @@
+  multiple network interfaces) they are said to be equivalent addresses.
+  Whenever we get an updated list of addresses from the resolver, we create an
+  entry in the map for each address, and populate the list of equivalent
+  addresses and EDS health-status. The subchannel is based on subchannel
+  creation/destruction or connection/disconnection from the child policy. The
+  equivalent address list computation on each resolver update is performed as
+  shown in the following pseudo-code:
+
+```
+// Resolver update contains a list of EquivalentAddressGroups,
+// each of which has a list of addresses and a health status.
+all_addresses = set()
+for eag in resolver_update:
+  for address in eag.addresses:
+    lb_policy.address_map[address].equivalent_addresses = eag.addresses
+    lb_policy.address_map[address].health_status = eag.health_status
+    all_addresses.insert(address)
+// Now remove equivalencies for any address not found in the current update.
+for address, entry in lb_policy.address_map:
+  if address not in all_addresses:
+    entry.equivalent_addresses.clear()
+```
+
+* The `xds_override_host_experimental` policy will always remove addresses in
+  state `DRAINING` when passing an update down to its child policy, regardless
+  of the value of `override_host_status`.
+
+* When the policy gets an update that changes an address's health status to
+  `DRAINING`, if (and only if) the `override_host_status` set includes
+  `DRAINING`, it will avoid closing the connection on that subchannel, even
+  when the child policy tries to do so. Instead, this policy will maintain the
+  connection until either (a) the connection closes due to the server closing
+  it or a network failure, or (b) the policy receives a subsequent EDS update
+  that completely removes the endpoint.
+
+* whenever a new subchannel is created (by the child policy that is
+  routing an RPC - see below), update the entry in `address_map` (for
+  the subchannel address i.e. the peer address as the key) with the new
+  subchannel value updated in the tuple
+  `[subchannel, health-status, equivalent addresses]`.
+
+    * in Java and Go, we may have to wait for subchannel to be `READY` in order
+      to know which specific address the subchannel is connected to before we
+      add it to the map.
+
+    * in C-core and Node, LB policies create a new subchannel for every address
+      every time there is a new address list. In the map, we will just replace
+      an existing entry with the most recent subchannel created for a given
+      address i.e. the new subchannel created will replace the older one for
+      that address.
+
+* whenever a subchannel is shut down, remove the associated entries from
+  `address_map`.
+
+    * this can be achieved by wrapping a subchannel to intercept subchannel
+      shutdown and the wrapper is returned by `createSubchannel`.
+
+* the policy's subchannel picker pseudo-code is as follows. `policy_config`
```

Comment:
We need to cover synchronization here to allow the picker to safely access the map that is owned by the control plane.

Reply:
done